beacon_node/http_api/src/beacon_proposer_cache.rs

```diff
@@ -0,0 +1,168 @@
+
+        // We *must* skip forward to the current epoch to obtain valid proposer
+        // duties. We cannot skip to the previous epoch, like we do with
+        // attester duties.
+        while head_state.current_epoch() < current_epoch {
+            // Skip slots until the current epoch, providing `Hash256::zero()` as the state root
+            // since we don't require it to be valid to identify producers.
+            per_slot_processing(&mut head_state, Some(Hash256::zero()), &chain.spec)?;
+        }
+
+        let proposers = current_epoch
+            .slot_iter(T::EthSpec::slots_per_epoch())
+            .map(|slot| {
+                head_state
+                    .get_beacon_proposer_index(slot, &chain.spec)
+                    .map_err(BeaconChainError::from)
+                    .and_then(|i| {
+                        let pubkey = chain
+                            .validator_pubkey(i)?
+                            .ok_or_else(|| BeaconChainError::ValidatorPubkeyCacheIncomplete(i))?;
+
+                        Ok(ProposerData {
+                            pubkey: PublicKeyBytes::from(pubkey),
+                            slot,
+                        })
+                    })
+            })
+            .collect::<Result<_, _>>()?;
+
+        let epoch_boundary_slot = head_state
+            .current_epoch()
+            .start_slot(T::EthSpec::slots_per_epoch());
+        let epoch_boundary_root = if head_state.slot >= epoch_boundary_slot {
+            head_root
+        } else {
+            *head_state.get_block_root(epoch_boundary_slot)?
+        };
+
+        Ok(Self {
+            epoch: current_epoch,
+            epoch_boundary_root,
+            proposers,
+        })
+    }
+
+    /// Return the proposers for the given `Epoch`.
+    ///
+    /// The cache may be rebuilt if:
+    ///
+    /// - The epoch has changed since the last cache build.
+    /// - There has been a re-org that crosses an epoch boundary.
+    pub fn get_proposers<T: BeaconChainTypes>(
+        &mut self,
+        chain: &BeaconChain<T>,
+        epoch: Epoch,
+    ) -> Result<Vec<ProposerData>, warp::Rejection> {
+        let is_prior_to_genesis = chain.slot_clock.is_prior_to_genesis().ok_or_else(|| {
+            warp_utils::reject::custom_server_error("unable to read slot clock".to_string())
+        })?;
+        let current_epoch = if is_prior_to_genesis {
+            chain.spec.genesis_slot.epoch(T::EthSpec::slots_per_epoch())
+        } else {
+            chain
+                .epoch()
+                .map_err(warp_utils::reject::beacon_chain_error)?
+        };
+
+        // Disallow requests that are outside the current epoch. This ensures the cache doesn't get
+        // washed-out with old values.
+        if current_epoch != epoch {
+            return Err(warp_utils::reject::custom_bad_request(format!(
+                "requested epoch is {} but only current epoch {} is allowed",
```

Comment:
So currently it's so that the value of the endpoint is easier to cache. We should really go back and add a slow, un-cached path for non-current requests, like we do for attester duties.

I've added this to #1434 so we can come back for it.